contracts/src/v0.8/functions/dev/1_0_0/FunctionsClient.sol

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import {Functions} from "./Functions.sol";
+import {IFunctionsRouter} from "./interfaces/IFunctionsRouter.sol";
+import {IFunctionsClient} from "./interfaces/IFunctionsClient.sol";
+import {IFunctionsCoordinator} from "./interfaces/IFunctionsCoordinator.sol";
+import {IFunctionsBilling} from "./interfaces/IFunctionsBilling.sol";
+
+/**
+ * @title The Chainlink Functions client contract
+ * @notice Contract writers can inherit this contract in order to create Chainlink Functions requests
+ */
+abstract contract FunctionsClient is IFunctionsClient {
+  IFunctionsRouter private s_router;
+  mapping(bytes32 => address) internal s_pendingRequests; /* requestId => fulfillment sender */
+
+  event RequestSent(bytes32 indexed id);
+  event RequestFulfilled(bytes32 indexed id);
+
+  error SenderIsNotRegistry();
+  error RequestIsAlreadyPending();
+  error RequestIsNotPending();
+
+  constructor(address router) {
+    setRouter(router);
+  }
+
+  /**
+   * @inheritdoc IFunctionsClient
+   */
+  function getDONPublicKey(bytes32 jobId) external view override returns (bytes memory) {
+    IFunctionsCoordinator coordinator = IFunctionsCoordinator(s_router.getRoute(jobId));
+    return coordinator.getDONPublicKey();
+  }
+
+  /**
+   * @notice Estimate the total cost that will be charged to a subscription to make a request: gas re-imbursement, plus DON fee, plus Registry fee
+   * @param req The initialized Functions.Request
+   * @param subscriptionId The subscription ID
+   * @param gasLimit gas limit for the fulfillment callback
+   * @return billedCost Cost in Juels (1e18) of LINK
+   */
+  function estimateCost(
```

Comment:
Like keep it to the tooling? So in the tooling get the Coordinator from the Router, then estimate?
Might make more sense that way especially because users won't really know which DON ID to use. That is handled by the tooling.